File: corpus/WikiMatrix/scripts/tsv2alg.py
#!/usr/bin/python3
#-*-python-*-
#
#

import sys
import argparse
from sqlitedict import SqliteDict

parser = argparse.ArgumentParser(description='Hash a corpus')
parser.add_argument('-s', type=str, default='source.db',
                    help='source language database')
parser.add_argument('-t', type=str, default='target.db',
                    help='target language database')
args = parser.parse_args()


srcdb = SqliteDict(args.s)
trgdb = SqliteDict(args.t)

count = 0
for line in sys.stdin:
    fields = line.rstrip().split("\t");
    if fields[1] in srcdb:
        if fields[2] in trgdb:
            sid = srcdb[fields[1]]
            tid = trgdb[fields[2]]
            print("\t".join((sid,tid,fields[0])), flush=True);
        else:
            print("not found (trg): " + fields[2], file=sys.stderr, flush=True)
    else:
        print("not found (src): " + fields[1], file=sys.stderr, flush=True)

            

    count += 1
    if (count % 10000 == 0):
        print('.', end = '', file=sys.stderr, flush=True)
    if (count % 500000 == 0):
        print(count, file=sys.stderr, flush=True)

srcdb.close()
trgdb.close()

File: corpus/WikiMatrix/scripts/tsv2alg.pl
#!/usr/bin/env perl
#-*-perl-*-
#
# look for sentence IDs in cdb databases
# make it possible to search several partitions


use utf8;
use strict;
use CDB_File;
use XML::Writer;

use vars qw($opt_s $opt_t $opt_S $opt_T);
use Getopt::Std;
getopts('s:t:S:T:');

my $SrcDbFile = $opt_s || "corpus.xx";
my $TrgDbFile = $opt_t || "corpus.yy";

my $SrcFile = $opt_S || "xx/corpus.xml.gz";
my $TrgFile = $opt_T || "yy/corpus.xml.gz";

my @SrcDBs = ();
my @TrgDBs = ();

&open_cdbs(\@SrcDBs,$SrcDbFile);
&open_cdbs(\@TrgDBs,$TrgDbFile);

my $XmlWriter = XML::Writer->new( DATA_MODE => 1,
				  DATA_INDENT => 1 );

$XmlWriter->xmlDecl("UTF-8");
$XmlWriter->doctype('cesAlign', "-//CES//DTD XML cesAlign//EN", "");
$XmlWriter->startTag('cesAlign','version' => '1.0');
$XmlWriter->startTag('linkGrp',
		     'targType' => 's',
		     'fromDoc'  => $SrcFile,
		     'toDoc'    => $TrgFile);


my $id = 0;
while (<>){
    chomp;
    my ($score,$src,$trg) = split(/\t/);
    if (my $sid = get_id(\@SrcDBs,$src)){
	if (my $tid = get_id(\@TrgDBs,$trg)){
	    # print join("\t",$sid,$tid,$score),"\n";
	    $id++;
	    $XmlWriter->emptyTag('link',
				 'id' => 'L'.$id,
				 'xtargets' => "$sid;$tid", 
				 'score' => $score);
	    print STDERR '.' if (! ($id % 2500));
	    print STDERR " $id\n" if (! ($id % 100000));
	}
	else{
	    print STDERR "target sentence '$trg' does not exist!\n";
	}
    }
    else{
	print STDERR "source sentence '$src' does not exist!\n";
    }
}

$XmlWriter->endTag("linkGrp");
$XmlWriter->endTag("cesAlign");


sub open_cdbs{
    my ($dbs,$filebase) = @_;
    my $file = $filebase;
    my $idx = 0;
    while (-e $file){
	$$dbs[$idx] = CDB_File->TIEHASH($file);
	$idx++;
	$file=$filebase.$idx;
    }
}

sub get_id{
    foreach my $db (@{$_[0]}){
	return $db->FETCH($_[1]) if ($db->EXISTS($_[1]));
    }
    return undef;
}
